File: go/ingest/encoding.go
// SPDX-License-Identifier: Apache-2.0

package ingest

import (
	"fmt"
	"strings"
	"unicode/utf8"
)

// detectEncoding inspects the leading bytes of raw for a BOM marker.
// Returns the decoded string, the encoding name, and any error.
func detectEncoding(raw []byte) (string, string, error) {
	// UTF-8 BOM (EF BB BF).
	if len(raw) >= 3 && raw[0] == 0xEF && raw[1] == 0xBB && raw[2] == 0xBF {
		stripped := raw[3:]
		if !utf8.Valid(stripped) {
			return "", "", fmt.Errorf("structured: invalid UTF-8 after BOM")
		}
		return string(stripped), "utf-8-bom", nil
	}

	// UTF-16 LE BOM (FF FE).
	if len(raw) >= 2 && raw[0] == 0xFF && raw[1] == 0xFE {
		decoded, err := decodeUTF16LE(raw[2:])
		if err != nil {
			return "", "", fmt.Errorf("structured: UTF-16 LE decode: %w", err)
		}
		return decoded, "utf-16-le", nil
	}

	// UTF-16 BE BOM (FE FF).
	if len(raw) >= 2 && raw[0] == 0xFE && raw[1] == 0xFF {
		decoded, err := decodeUTF16BE(raw[2:])
		if err != nil {
			return "", "", fmt.Errorf("structured: UTF-16 BE decode: %w", err)
		}
		return decoded, "utf-16-be", nil
	}

	// Try UTF-8.
	if utf8.Valid(raw) {
		return string(raw), "utf-8", nil
	}

	// Fallback: Latin-1 (ISO-8859-1) — every byte is valid.
	var b strings.Builder
	b.Grow(len(raw))
	for _, by := range raw {
		b.WriteRune(rune(by))
	}
	return b.String(), "latin-1", nil
}

// decodeUTF16LE decodes UTF-16 little-endian bytes to a Go string.
func decodeUTF16LE(data []byte) (string, error) {
	if len(data)%2 != 0 {
		return "", fmt.Errorf("odd byte count for UTF-16")
	}
	var b strings.Builder
	b.Grow(len(data) / 2)
	for i := 0; i+1 < len(data); i += 2 {
		cp := rune(data[i]) | rune(data[i+1])<<8
		b.WriteRune(cp)
	}
	return b.String(), nil
}

// decodeUTF16BE decodes UTF-16 big-endian bytes to a Go string.
func decodeUTF16BE(data []byte) (string, error) {
	if len(data)%2 != 0 {
		return "", fmt.Errorf("odd byte count for UTF-16")
	}
	var b strings.Builder
	b.Grow(len(data) / 2)
	for i := 0; i+1 < len(data); i += 2 {
		cp := rune(data[i])<<8 | rune(data[i+1])
		b.WriteRune(cp)
	}
	return b.String(), nil
}

File: sdks/ts/memory/src/ingest/encoding.ts
// SPDX-License-Identifier: Apache-2.0

/**
 * Character encoding detection for structured data extractors.
 * Detects UTF-8 (with/without BOM), UTF-16 LE/BE, and Latin-1
 * (ISO-8859-1) fallback.
 */

export const detectEncoding = (raw: Buffer): { text: string; encoding: string } => {
  // UTF-8 BOM (EF BB BF).
  if (raw.length >= 3 && raw[0] === 0xef && raw[1] === 0xbb && raw[2] === 0xbf) {
    return { text: raw.subarray(3).toString('utf8'), encoding: 'utf-8-bom' }
  }

  // UTF-16 LE BOM (FF FE).
  if (raw.length >= 2 && raw[0] === 0xff && raw[1] === 0xfe) {
    return { text: raw.subarray(2).toString('utf16le'), encoding: 'utf-16-le' }
  }

  // UTF-16 BE BOM (FE FF).
  if (raw.length >= 2 && raw[0] === 0xfe && raw[1] === 0xff) {
    return { text: decodeUTF16BE(raw.subarray(2)), encoding: 'utf-16-be' }
  }

  // Try UTF-8. If all bytes are valid, use it.
  const utf8Text = raw.toString('utf8')
  const roundTrip = Buffer.from(utf8Text, 'utf8')
  if (roundTrip.length === raw.length && raw.equals(roundTrip)) {
    return { text: utf8Text, encoding: 'utf-8' }
  }

  // Fallback: Latin-1 (ISO-8859-1).
  return { text: raw.toString('latin1'), encoding: 'latin-1' }
}

const decodeUTF16BE = (data: Buffer): string => {
  const swapped = Buffer.alloc(data.length)
  for (let i = 0; i + 1 < data.length; i += 2) {
    swapped[i] = data[i + 1] ?? 0
    swapped[i + 1] = data[i] ?? 0
  }
  return swapped.toString('utf16le')
}
